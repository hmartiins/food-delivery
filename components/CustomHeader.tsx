import { Image, Text, TouchableOpacity, View } from 'react-native';

import { useRouter } from 'expo-router';

import { images } from '@/constants';
import { CustomHeaderProps } from '@/type';

export const CustomHeader = ({ title }: CustomHeaderProps) => {
  const router = useRouter();

  return (
    <View className="custom-header">
      <TouchableOpacity onPress={() => router.back()} testID="back-button">
        <Image
          source={images.arrowBack}
          className="size-5"
          resizeMode="contain"
        />
      </TouchableOpacity>

      {title && <Text className="base-semibold text-dark-100">{title}</Text>}

      <TouchableOpacity
        onPress={() => router.push('/search')}
        testID="search-button"
      >
        <Image
          source={images.search}
          className="size-5"
          resizeMode="contain"
          testID="search-icon"
        />
      </TouchableOpacity>
    </View>
  );
};
