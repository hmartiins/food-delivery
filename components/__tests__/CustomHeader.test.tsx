import { fireEvent, render } from '@testing-library/react-native';

import { CustomHeader } from '../CustomHeader';

const mockBack = jest.fn();
const mockPush = jest.fn();

jest.mock('expo-router', () => ({
  useRouter: () => ({
    back: mockBack,
    push: mockPush,
  }),
}));

jest.mock('@/constants', () => ({
  images: {
    arrowBack: 'arrow-back-icon',
    search: 'search-icon',
  },
}));

describe('<CustomHeader />', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('renders correctly without title', () => {
    const { getByTestId, root } = render(<CustomHeader />);

    const textElements = root.findAllByType('Text');
    expect(textElements).toHaveLength(0);

    expect(getByTestId('back-button')).toBeTruthy();
    expect(getByTestId('search-button')).toBeTruthy();
  });

  test('renders correctly with title', () => {
    const { getByText } = render(<CustomHeader title="My Title" />);
    expect(getByText('My Title')).toBeTruthy();
  });

  test('calls router.back when back button is pressed', () => {
    const { getByTestId } = render(<CustomHeader title="Test" />);
    const backButton = getByTestId('back-button');

    fireEvent.press(backButton);

    expect(mockBack).toHaveBeenCalledTimes(1);
    expect(mockPush).not.toHaveBeenCalled();
  });

  test('calls router.push with /search when search button is pressed', () => {
    const { getByTestId } = render(<CustomHeader title="Test" />);
    const searchButton = getByTestId('search-button');

    fireEvent.press(searchButton);

    expect(mockPush).toHaveBeenCalledWith('/search');
    expect(mockPush).toHaveBeenCalledTimes(1);
    expect(mockBack).not.toHaveBeenCalled();
  });

  test('handles empty string title correctly', () => {
    const { queryByText } = render(<CustomHeader title="" />);
    expect(queryByText('')).toBeNull();
  });

  test('snapshot of the component', () => {
    const { toJSON } = render(<CustomHeader title="Header Snapshot Test" />);
    expect(toJSON()).toMatchSnapshot();
  });
});
