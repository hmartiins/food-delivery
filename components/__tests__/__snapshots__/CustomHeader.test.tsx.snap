// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CustomHeader /> snapshot of the component 1`] = `
<View
  className="custom-header"
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="back-button"
  >
    <Image
      className="size-5"
      resizeMode="contain"
      source="arrow-back-icon"
    />
  </View>
  <Text
    className="base-semibold text-dark-100"
  >
    Header Snapshot Test
  </Text>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="search-button"
  >
    <Image
      className="size-5"
      resizeMode="contain"
      source="search-icon"
      testID="search-icon"
    />
  </View>
</View>
`;
